doc/src/core-transmutation.md

```diff
@@ -0,0 +1,143 @@
+Challenge: Preconditions for core transmuting methods
+
+## Status
+
+\[Open | Resolved | Expired\]: Open
+
+## Dates
+
+Start date:
+
+End date:
+
+## Goal
+
+Confirm the correctness of value transmutations performed by libcore, including those transmutations exposed as public methods provided by libcore.
+
+To keep the goal somewhat manageable, it excludes some classes of code (e.g. utf8-validation, async tasks, and others); see the assumptions for the full list of excluded categories.
+
+## Details
+
+### Motivating Example
+
+There are several calls to these unsafe methods within the code of safe methods exported libcore itself, so having clear and verifiable safety contracts is critical for verifying the safety of the safe methods that invoke them.
+
+For example, `std::slice::align_to` (which unsafely converts any slice `&[T]` into a tuple `(&[T], &[U], &[T])` composed of a prefix, a maximal sequence of values of type `U`, and a suffix) just says for its safety that “all the usual caveats pertaining to `transmute::<T, U>` also apply here”, but the documentation for `transmute` is frankly (and rightfully) scary.
+
+### Part 1: the two intrinsics
+
+The public unsafe intrinsic `transmute<T, U>` takes a value of type `T` and reinterprets it to have type `U`, with the sole (statically-enforced) restriction being that the types `T` and `U` must have the same size. The unstable intrinsic `transmute_unchecked` is similar, except that it removes the static size restriction, treating violations of it as undefined behavior.
+
+What is the right way to encode the preconditions of the two transmutation intrinsics?
+
+If one is solely concerned about safety: The Rustonomicon lists several ways that transmutation can yield undefined behavior. Encoding a correct safety contract for transmute is crucial.
+
+If one is concerned about proving functional correctness, then reasoning about transmutation will require reasoning about the byte representation of values to justify that the reinterpretation of the value’s bytes for satisfying the pending proof obligation associated with the output value.
```

Comment:
Proving the functional correctness of what ? isn't transmute safety already about byte level representations ?

Reply:
@pnkfelix could you provide more context?

Reply:
To prove safety, one must ensure the following conditions:

1. Size compatibility: The source and destination types must have the same size. Transmuting types of different sizes can lead to memory corruption.
2. Alignment: The alignment requirements of the target type must be met.
3. Valid bit pattern: The bit pattern of the source type must be a valid bit pattern for the target type.
4. No uninitialized memory: Transmuting to a type with stricter requirements (e.g., from integer to reference) can cause undefined behavior if the data is not properly initialized.

What are the extra conditions for functional correctness?

Reply:
Here’s the kind of thing I was thinking of when I wrote the above text about full functional correctness:  if you are transmuting from a type T to a type U, *and* the type U has some invariant associated with it, then chances are that you might need to actually provide a chain of inductive reasoning that says that T itself has some associated invariant that will provide the necessary implication of the invariant associated with U.

To be clear: I recognize that the above may not be true in general. I can imagine cases where one does a transmute and then has some further “fixup” that needs to occur to establish library invariants, or other extra conditions on other data, unrelated to T, that are necessary for establishing the invariant associated with U.

Anyway, the point is that from my perspective, full functional correctness is going to be deeply tied to the customer’s specification of what they need from the given data abstraction. So of course its going to require deeper levels of reasoning that what you would need solely for language level validity invariants.